tests/test/scheduler/test_multiple_mpi_worlds.cpp

```diff
@@ -0,0 +1,291 @@
+    const std::string actualHostB = worldB.getHostForRank(0);
+    REQUIRE(actualHostA == faabric::util::getSystemConfig().endpointHost);
+    REQUIRE(actualHostB == faabric::util::getSystemConfig().endpointHost);
+
+    worldA.destroy();
+    worldB.destroy();
+}
+
+TEST_CASE_METHOD(MultiWorldMpiTestFixture,
+                 "Test send and recv on same host from multiple worlds",
+                 "[mpi]")
+{
+    // Send a message between colocated ranks
+    int rankA1 = 0;
+    int rankA2 = 1;
+    std::vector<int> messageData = { 0, 1, 2 };
+    worldA.send(
+      rankA1, rankA2, BYTES(messageData.data()), MPI_INT, messageData.size());
+    worldB.send(
+      rankA1, rankA2, BYTES(messageData.data()), MPI_INT, messageData.size());
+
+    SECTION("Test queueing")
+    {
+        // Check for world A
+        REQUIRE(worldA.getLocalQueueSize(rankA1, rankA2) == 1);
+        REQUIRE(worldA.getLocalQueueSize(rankA2, rankA1) == 0);
+        REQUIRE(worldA.getLocalQueueSize(rankA1, 0) == 0);
+        REQUIRE(worldA.getLocalQueueSize(rankA2, 0) == 0);
+        const std::shared_ptr<InMemoryMpiQueue>& queueA2 =
+          worldA.getLocalQueue(rankA1, rankA2);
+        faabric::MPIMessage actualMessage = *(queueA2->dequeue());
+        // checkMessage(actualMessage, worldId, rankA1, rankA2, messageData);
+
+        // Check for world B
+        REQUIRE(worldB.getLocalQueueSize(rankA1, rankA2) == 1);
+        REQUIRE(worldB.getLocalQueueSize(rankA2, rankA1) == 0);
+        REQUIRE(worldB.getLocalQueueSize(rankA1, 0) == 0);
+        REQUIRE(worldB.getLocalQueueSize(rankA2, 0) == 0);
+        const std::shared_ptr<InMemoryMpiQueue>& queueA2B =
+          worldB.getLocalQueue(rankA1, rankA2);
+        actualMessage = *(queueA2B->dequeue());
+        // checkMessage(actualMessage, worldId, rankA1, rankA2, messageData);
+    }
+
+    SECTION("Test recv")
+    {
+        MPI_Status status{};
+        auto bufferAllocation = std::make_unique<int[]>(messageData.size());
+        auto* buffer = bufferAllocation.get();
+
+        // Check for world A
+        worldA.recv(
+          rankA1, rankA2, BYTES(buffer), MPI_INT, messageData.size(), &status);
+        std::vector<int> actual(buffer, buffer + messageData.size());
+        REQUIRE(actual == messageData);
+        REQUIRE(status.MPI_ERROR == MPI_SUCCESS);
+        REQUIRE(status.MPI_SOURCE == rankA1);
+        REQUIRE(status.bytesSize == messageData.size() * sizeof(int));
+
+        // Check for world B
+        worldB.recv(
+          rankA1, rankA2, BYTES(buffer), MPI_INT, messageData.size(), &status);
+        std::vector<int> actualB(buffer, buffer + messageData.size());
+        REQUIRE(actualB == messageData);
+        REQUIRE(status.MPI_ERROR == MPI_SUCCESS);
+        REQUIRE(status.MPI_SOURCE == rankA1);
+        REQUIRE(status.bytesSize == messageData.size() * sizeof(int));
+    }
+}
+
+TEST_CASE_METHOD(RemoteMultiWorldMpiTestFixture,
```

Comment:
This is the test that fails. Unfortunately, as explained in the PR, I don't think there is a way to properly test this without actually porting the MPI-native tests to our distributed tests.